man/man1/create_bmp_for_stripline_coupler.1
.TH create_bmp_for_stripline_coupler 1 "atlc-.4.5.0 28th Sept 2003" "Dr. David Kirkby"
.ds n 5
.SH NAME
\fBcreate_bmp_for_stripline_coupler\fR - bitmap generator for coupler with thin striplines between two infinitely wide groundplanes (part of \fBatlc\fR)
.SH SYNOPSIS
\fBcreate_bmp_for_stripline_coupler [-C] [-b bitmapsize] [-v] H w s Er outputfile\fR
.br
.SH WARNING
This man page is not a complete set of documentation - the complexity of the atlc project makes man pages not an ideal way to document it, although out of completeness, man pages are produced. 
The best documentation that was current at the time the version was produced should be found on your
hard drive, usually at 
.br
/usr/local/share/atlc/docs/html\-docs/index.html 
.br
although it might be elsewhere if
your system administrator chose to install the package elsewhere. Sometimes, errors are corrected
in the documentation and placed at http://atlc.sourceforge.net/ before a new release of atlc is
released.  Please, if you notice a problem with the documentation - even spelling errors and typos,
please let me know. 

.SH DESCRIPTION
\fBcreate_bmp_for_stripline_coupler\fR is a pre-processor for \fBatlc\fR, the finite difference 
program that is used to calculate the 
properties of a two and three conductor electrical transmission lines of arbitrary 
cross section. The program \fBcreate_bmp_for_stripline_coupler\fR is used as a fast way of
generating bitmaps (there is no need to use a graphics program), for a
pair of 1 pixel thick conductors of width w, spaced a distance s apart, 
inside a pair of
groundplanes of  separation H. This has an exact analytical
answer, when the inner is infinitely thin and the overall width W is infinitely
wide. This theoretical impedance is calculated and printed to stderr if:
.br
a) The -v option is given on create_bmp_for_stripline_coupler's command line. AND
.br
b) create_bmp_for_stripline_coupler was linked against the gsl library at http://sources.redhat.com/gsl/
.P
-----------------------------------------------------  ^
.br
|                                                   |  |
.br                                                     
|                  Er                               |  |
.br                                                     
|                                                   |  |
.br                                                     
|            -----------       -----------          |  H
.br
|            <----w----><--s--><----w---->          |  |
.br       
|                                                   |  |
.br
|                                                   |  |
.br
|                                                   |  |
.br
-----------------------------------------------------  v
.br
<-------------------------W------------------------->
.br

The overall width W is calculated internally to be large enough that
making it any larger will have negligible effect on the results.
The dielectric is assumed that of vacuum.
.PP 
.br
The bitmaps produced by \fBcreate_bmp_for_stripline_coupler\fR are 24-bit bit colour bitmaps, as are required by \fBatlc\fR.

.SH OPTIONS
.PP
\fB-C \fR
.br
Print the full copyright information and exit.
.PP
.br
\fB-b bitmapsize\fR
.br
is used to set the size of the bitmap, and so the accuracy to which atlc
is able to calculate the coupler's properties. The default
value for 'bitmapsize' is normally 18 although this can be set at
compile time. (see 'DEFAULT_COUPLER_BITMAP_SIZE' in definitions.h). The value can be set anywhere from 6 to 28. The size of bitmap produced will be
between 2^(bitmapsize-1) and 2^(bitmapsize+1), so if its set to 28 the
bitmap could be as big as 512 Mb, which would take an eternity to
compute. 
.PP
\fB-v \fR
.br
Print some extra data to stdout on the theoretical impedance of the
structure. This is useful to compare with the numerical result from
\fBatlc\fR
.SH EXAMPLES
.br
In the first example, a standard coupler is created, where the 


eg1. create_bmp_for_stripline_coupler  1.23 0.7 1.2 1.0  coupler.bmp
.br
eg2. create_bmp_for_stripline_coupler -b 8 1.23 0.7 3.3  small.bmp 
.br
eg3. create_bmp_for_stripline_coupler -v 12 1.23 0.7 4.8  theory.bmp  
.br













.P
.SH SEE ALSO
atlc(1)
create_bmp_for_circ_in_circ(1)
create_bmp_for_circ_in_rect(1)
create_bmp_for_microstrip_coupler(1)
create_bmp_for_rect_cen_in_rect(1)
create_bmp_for_rect_cen_in_rect_coupler(1)
create_bmp_for_rect_in_circ(1)
create_bmp_for_rect_in_rect(1)
create_bmp_for_symmetrical_stripline(1)
design_coupler(1)
find_optimal_dimensions_for_microstrip_coupler(1)
readbin(1)
.P 
.br
http://atlc.sourceforge.net                - Home page 
.br
http://sourceforge.net/projects/atlc       - Download area
.br
atlc-X.Y.Z/docs/html-docs/index.html       - HTML docs
.br
atlc-X.Y.Z/docs/qex-december-1996/atlc.pdf - theory paper
.br
atlc-X.Y.Z/examples                        - examples
.br
.SH AUTHOR
Dr. David Kirkby PhD, drkirkby at ntlworld.com 
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2002 Dr. David Kirkby. 
.PP

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.
.pp
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.p
You should have received a copy of the GNU General Public License (in
the file COPYING) along with this program. You may get a full copy of it
by running the program with the -C option. Otherwise you can write to the 
Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, 
MA  02111-1307, USA.


NEWS
NEWS for realease 4.3.0 on 23 May 2003.

A bug, preventing atlc from runningt properly on multiple
processors has been fixed. This is the main change. 

atlc should no longer fail any test on a multi-processor
system configured with --with-threads. The algorithm used
for both single processor and multiple processor (i.e. 
threaded code is the same). 

The -t option to atlc, which sets the number of threads 
when configured for multiple processors can be set to 0
to use entirely the single-threaded algorithm. If set to
one, it will use the multi-threaded algoritm, but use
only one thread. If set to some other number, it will use
that number of threads and be optimal for the same number
of cpus. 

Support for distributed programming has been temporarily removed.
Someone else added the code, but since he is no longer involved in
the development, it is difficult for me to follow. So I'll remove
the MPI specific code and will re-implement this. 

A test has been added to make sure atlc can handle the -d option 


Dr. David Kirkby 23/5/2003



src/non_gui/create_bmp_for_stripline_coupler.c
/* atlc - arbitrary transmission line calculator, for the analysis of
transmission lines are directional couplers. 

Copyright (C) 2002. Dr. David Kirkby, PhD (G8WRB).

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either package_version 2
of the License, or (at your option) any later package_version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
USA.

Dr. David Kirkby, e-mail drkirkby@ntlworld.com 

*/

/* The following calculates the odd and even mode impedances between two
zero thickness strips of width w, spaced a distance s between two
groundplanes of spacing h. */
#include "config.h"


#define RATIO 8
#include "definitions.h"
#include "exit_codes.h"

#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif

#ifdef HAVE_STRING_H
#include <string.h>
#endif

#ifdef HAVE_GSL_GSL_SF_ELLINT_H
#include <gsl/gsl_sf_ellint.h>
#endif

double WW, HH, ww, Er1, Er2;
static double ss; 
int W, H, s,  w,  verbose=FALSE;

int main(int argc, char **argv)
{
  double er, return_error;
  double Zodd, Zeven, Zo;
  int bmp_size=DEFAULT_COUPLER_BMP_SIZE, q; 
  FILE *image_data_fp;
  struct transmission_line_properties optimise;
  struct transmission_line_properties not_used;

  not_used.W=1; /* keep MipsPro compiler happy on IRIX */

  while((q=get_options(argc,argv,"Cb:v")) != -1)
  switch (q) 
  {
    case 'C':
      print_copyright((char *) "2002");
      exit_with_msg_and_exit_code("",OKAY);
    break;
    case 'b':
      bmp_size=atol(my_optarg); 
    break;
    case 'v':
      verbose++;
    break;
  }
  if( bmp_size < 6 || (argc-my_optind != 5) || bmp_size>28)
  {
    usage_create_bmp_for_stripline_coupler();
    exit_with_msg_and_exit_code("",PROGRAM_CALLED_WITH_WRONG_NUMBER_OF_ARGUMENTS);
  }
  HH=atof(argv[my_optind]);
  ww=atof(argv[my_optind+1]);
  ss=atof(argv[my_optind+2]);
  if(HH<0 || ww <0 || ss<0)
  {
    fprintf(stderr,"Sorry, W, H and s must all be greater than 0\n");
    exit_with_msg_and_exit_code("",DIMENSION_LESS_THAN_ZERO);
  }
  er=atof(argv[my_optind+3]);
  Er1=er;
  Er2=er;
  if(er < 1.0)
  {
    fprintf(stderr,"Sorry, you can't have a dielectric constand Er of less than 1.0\n");
    exit_with_msg_and_exit_code("Sorry, you can't have a dielectric constand Er of less than 1.0",PERMITTIVITY_LESS_THAN_1);
  }
  if( (image_data_fp=fopen(argv[my_optind+4],"wb")) ==NULL)
    exit_with_msg_and_exit_code("Can't open binary file for writing",CANT_OPEN_FOR_WRITING);
  WW=2.0*ww+ss+RATIO*HH;
  optimise.float_values[0]=WW            ;   /* minimum width as a float*/
  optimise.float_values[1]=HH;               /* height in floats */
  optimise.float_values[2]=ww;               /* stripline width */
  optimise.float_values[3]=ss;               /* spacing between the strips */

  optimise.importance[0]=NOT_IMPORTANT;      /* W is non critical */
  optimise.importance[1]=MOST_IMPORTANT;          /* H is critical */
  optimise.importance[2]=IMPORTANT;          /* w is critical */
  optimise.importance[3]=IMPORTANT;     /* s is most critical */

  optimise.odd_or_even[0]=DONT_CARE;        /* W can be odd or even */
  optimise.odd_or_even[1]=ODD;              /* H must be even */
  optimise.odd_or_even[2]=DONT_CARE;        /* w can be odd or even */
  optimise.odd_or_even[3]=DONT_CARE;        /* s can be odd or even */

  /* We will now optimise for the 4 parameters W, H, w and s, forcing H
  to be odd, and not bothering what the width is, as long as its about
  right. */

  return_error=calculate_integer_values(&optimise, 4, bmp_size);
  if(verbose >=2)
    printf("error returned from calculate_integer_values=%g\n", return_error);
  W=optimise.best[0];
  H=optimise.best[1];
  w=optimise.best[2];
  s=optimise.best[3];
  write_bitmap(image_data_fp, not_used);
  if(verbose >= 1)
  {
    calculate_Zodd_and_Zeven(&Zodd, &Zeven, &Zo, ww, HH, ss, er);
    printf("The actual dimensions you gave have theoretical imedances of:\n");
    printf("        Zodd= %f Zeven= %f Zo= %f (Ohms) ww=%f HH=%f ss=%f er=%f\n\n", Zodd, Zeven, Zo,ww,HH,ss,er);
    calculate_Zodd_and_Zeven(&Zodd, &Zeven, &Zo, (double) w, (double) H-10, (double) s, er);
    printf("The bitmap produced (which approximates what you want) should have:\n");
    printf("   Zodd= %f Zeven= %f Zo= %f (Ohms)\n", Zodd, Zeven, Zo);
  }
  return(OKAY);
}
